File: 1466-reorder-routes-to-make-all-paths-lead-to-the-city-zero/1466-reorder-routes-to-make-all-paths-lead-to-the-city-zero.java
class Solution {
    public void dfs(List<List<Pair<Integer, Integer>>> adj, boolean[] visited, int[] minChange, int currCity) {
        visited[currCity] = true;
        for (Pair<Integer, Integer> neighbourCity : adj.get(currCity)) {
            if (!visited[neighbourCity.getKey()]) {
                if (neighbourCity.getValue() == 1) {
                    minChange[0]++;
                }
                dfs(adj, visited, minChange, neighbourCity.getKey());
            }
        }
    }
    
    public int minReorder(int n, int[][] connections) {
        List<List<Pair<Integer, Integer>>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
        }
        for (int[] connection : connections) {
            adj.get(connection[0]).add(new Pair<>(connection[1], 1));
            adj.get(connection[1]).add(new Pair<>(connection[0], -1));
        }
        boolean[] visited = new boolean[n];
        int[] minChange = new int[1];
        dfs(adj, visited, minChange, 0);
        return minChange[0];
    }
}

File: LeetCode/1466-reorder-routes-to-make-all-paths-lead-to-the-city-zero/1466-reorder-routes-to-make-all-paths-lead-to-the-city-zero.cpp
class Solution {
public:
    int count = 0;
    
    void dfs(unordered_map<int, vector<pair<int, int>>> &adj, int source, int parent) {
        for (auto &p : adj[source]) {
            int v = p.first;
            int c = p.second;
            if (v != parent) {
                if (c == 1) {
                    count++;
                }
                dfs(adj, v, source);
            }
        }
    }
    
    int minReorder(int n, vector<vector<int>>& connections) {
        unordered_map<int, vector<pair<int, int>>> adj;
        
        for (vector<int>& con : connections) {
            adj[con[0]].push_back({con[1], 1});
            adj[con[1]].push_back({con[0], 0});
        }
        
        dfs(adj, 0, -1);
        return count;
    }
};